src/app/qgsapplayertreeviewmenuprovider.cpp

```diff
@@ -328,16 +328,39 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
         if ( !layer->isInScaleRange( mCanvas->scale() ) )
           menu->addAction( tr( "Zoom to &Visible Scale" ), QgisApp::instance(), &QgisApp::zoomToLayerScale );
 
-        QMenu *menuSetCRS = new QMenu( tr( "Set CRS" ), menu );
-        // set layer crs
-        QAction *actionSetLayerCrs = new QAction( tr( "Set Layer CRS…" ), menuSetCRS );
-        connect( actionSetLayerCrs, &QAction::triggered, QgisApp::instance(), &QgisApp::setLayerCrs );
-        menuSetCRS->addAction( actionSetLayerCrs );
+        QMenu *menuSetCRS = new QMenu( tr( "Layer CRS" ), menu );
+        QAction *actionCurrentCrs = new QAction( layer->crs().authid(), menuSetCRS );
+        actionCurrentCrs->setEnabled( false );
+        menuSetCRS->addAction( actionCurrentCrs );
         // assign layer crs to project
         QAction *actionSetProjectCrs = new QAction( tr( "Set &Project CRS from Layer" ), menuSetCRS );
         connect( actionSetProjectCrs, &QAction::triggered, QgisApp::instance(), &QgisApp::setProjectCrsFromLayer );
         menuSetCRS->addAction( actionSetProjectCrs );
 
+        const QList< QgsCoordinateReferenceSystem> recentProjections = QgsCoordinateReferenceSystem::recentCoordinateReferenceSystems();
+        if ( !recentProjections.isEmpty() )
+        {
+          menuSetCRS->addSeparator();
+          int i = 0;
+          for ( const QgsCoordinateReferenceSystem &crs : recentProjections )
+          {
+            QAction *action = menuSetCRS->addAction( crs.userFriendlyIdentifier() );
+            action->setProperty( "layerId", layer->id() );
+            action->setProperty( "crs", crs.authid() );
```

Comment:
Not all crs have an auth id. The only reliable way to handle this would be to export the crs to wkt and use this as the data here.